core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java

```diff
@@ -0,0 +1,182 @@
+ *
+ * @author Pedro Ruivo
+ * @since 11.0
+ */
+@Test(groups = "functional")
+public abstract class BaseIracPersistenceTest<V> extends SingleCacheManagerTest {
+
+   private static final AtomicLong V_GENERATOR = new AtomicLong();
+   private static final String SITE = "LON";
+   protected String tmpDirectory;
+   protected AdvancedLoadWriteStore<String, V> cacheStore;
+   protected MarshallableEntryFactory<String, V> entryFactory;
+
+   public void testWriteAndLoad(Method method) {
+      String key = TestingUtil.k(method);
+      String value = TestingUtil.v(method);
+      IracMetadata metadata = createMetadata();
+
+      cacheStore.write(createEntry(key, value, metadata));
+
+      MarshallableEntry<String, V> loadedMEntry = cacheStore.loadEntry(key);
+
+
+      assertCorrectEntry(loadedMEntry, key, value, metadata);
+   }
+
+   public void testWriteAndPublisher(Method method) {
+      String key = TestingUtil.k(method);
+      String value = TestingUtil.v(method);
+      IracMetadata metadata = createMetadata();
+
+      cacheStore.write(createEntry(key, value, metadata));
+
+      MarshallableEntrySubscriber<V> subscriber = new MarshallableEntrySubscriber<>();
+      cacheStore.entryPublisher(key::equals, true, true).subscribe(subscriber);
+
+
+      List<MarshallableEntry<String, V>> entries = subscriber.cf.join();
+
+      AssertJUnit.assertEquals(1, entries.size());
+      assertCorrectEntry(entries.get(0), key, value, metadata);
+   }
+
+   @BeforeClass(alwaysRun = true)
+   @Override
+   protected void createBeforeClass() throws Exception {
+      tmpDirectory = CommonsTestingUtil.tmpDirectory(this.getClass());
+      Util.recursiveFileRemove(tmpDirectory);
+      boolean created = new File(tmpDirectory).mkdirs();
+      log.debugf("Created temporary directory %s (exists? %s)", tmpDirectory, !created);
+      super.createBeforeClass();
+   }
+
+   @AfterClass(alwaysRun = true)
+   @Override
+   protected void destroyAfterClass() {
+      super.destroyAfterClass();
+      Util.recursiveFileRemove(tmpDirectory);
+   }
+
+   @Override
+   protected EmbeddedCacheManager createCacheManager() throws Exception {
+      GlobalConfigurationBuilder gBuilder = createGlobalConfigurationBuilder();
+      ConfigurationBuilder cBuilder = new ConfigurationBuilder();
+      addPersistence(cBuilder);
+      EmbeddedCacheManager cm = TestCacheManagerFactory.createCacheManager(gBuilder, cBuilder);
+      cacheStore = TestingUtil.getFirstLoader(cm.getCache());
+      //noinspection unchecked
+      entryFactory = TestingUtil.extractComponent(cm.getCache(), MarshallableEntryFactory.class);
+      return cm;
+   }
+
+   protected SerializationContextInitializer getSerializationContextInitializer() {
+      return TestDataSCI.INSTANCE;
+   }
+
+   protected abstract void addPersistence(ConfigurationBuilder builder);
+
+   protected abstract V wrap(String key, String value);
```

Comment:
The `wrap` and `unwrap` implementations always return the value except in the case of the `IracJpaStoreTest`. We should create a default impl of:

```java
   @Override
   protected String wrap(String key, String value) {
      return value;
   }

   @Override
   protected String unwrap(String value) {
      return value;
   }
```

and override it in `IracJpaStoreTest`.

Reply:
-1 
That would require to remove the type `V` and use casting (and unchecked casts)
I can use `org.infinispan.persistence.PersistenceCompatibilityTest.KeyValueWrapper<K,V,T>`  since I did hit the same "problem" before.